docs/hugo/content/design/reconcile-interface.md

```diff
@@ -0,0 +1,227 @@
+    }
+
+    // Take the action needed
+    var result ctrl.Result
+    var err error
+    if !metaObj.GetDeletionTimestamp().IsZero() {
+        // Check the reconcile policy to ensure we're allowed to issue a delete
+        reconcilePolicy := GetReconcilePolicy(metaObj, log)
+        if !reconcilePolicy.AllowsDelete() {
+            log.V(Info).Info("Bypassing delete of resource due to policy", "policy", reconcilePolicy)
+            return ctrl.Result{}, RemoveResourceFinalizer(ctx, log, metaObj)
+        }
+
+        // Check if we actually need to issue a delete
+        hasFinalizer := controllerutil.ContainsFinalizer(metaObj, reconcilers.GenericControllerFinalizer)
+        if !hasFinalizer {
+            log.Info("Deleted resource")
+            return ctrl.Result{}, nil
+        }
+        result, err = r.Delete(ctx, log, gr.Recorder, metaObj) // It would be the responsibility of the reconcilers delete to check anything specific that it did during Claim and short-circuit if needed
+    } else {
+        // Check the reconcile policy to ensure we're allowed to issue a create or update
+        reconcilePolicy := reconcilers.GetReconcilePolicy(metaObj, log)
+        if !reconcilePolicy.AllowsModify() {
+            return handleSkipReconcile(ctx, log, metaObj) // This would call GetStatus to optionally fill out the status and Commit that
+        }
+
+        result, err = r.CreateOrUpdate(ctx, log, gr.Recorder, metaObj)
+    }
+
+    if readyErr, ok := conditions.AsReadyConditionImpactingError(err); ok {
+        log.Error(readyErr, "Encountered error impacting Ready condition")
+        err = gr.WriteReadyConditionError(ctx, metaObj, readyErr)
+    }
+    if err != nil {
+        log.Error(err, "Error claiming resource")
+        return ctrl.Result{}, reconcilers.IgnoreNotFoundAndConflict(err)
+    }
+```
+
+
+# FAQ
+
+**Q:** What component would own commiting the object to etcd?
+**A:** `GenericController` would always do a commit to etcd after a reconcile. It may need to check that the actual individual reconciler changed `Obj`. If no change it won't commit. Commit ability will not be taken from individual reconcilers though, if 
+       they need it they can do extra commits using the `KubeClient` they have access to.
+
+**Q:** What component would own setting the ready condition?
+**A:** The `GenericController`. Custom errors would be propagated via the error mechanism we have now from the individual reconciler to the `GenericController`. 
+       There are no custom success states currently, but if there were we wouldn't block the reconciler from setting them manually if we wanted some in the future.
+
+**Q:** Why split out `Claim`?
+**A:** Part of `Claim` is general (adding the standard finalizer). Part of it is reconciler specific (for example checking if the owner exists). `Claim` is part of a fault-tolerance model. The resource must actually succeed in writing its finalizer to
```

Comment:
Ah, ok - I think this answers my question above, it's things like checking the owner exists (but not interacting with ARM/DB).

Reply:
Yeah exactly. I've improved the wording some to clarify this but basically the idea with `Claim` is "perform what checks you need to perform before actually taking an action". So that's things like:

- Add finalizer (generic, can be lifted up into the generic controller).
- Check that owner exists (specific to resource types that have owners, which is everything right now but may not be everything going forward).
- Calculate the ARM ID of the resource we're about to create or update (for ARM resources only).

So the idea would be:
// Do resource specific Claim actions
// if err != nil, bail out here
// Do generic claim actions (finalizer)
// Do CreateOrUpdate